cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +211,379 @@ impl UpdateConfig for Api {
     }
 }
 
-fn load_inner(path: &std::path::Path) -> Result<Option<Vec<u8>>, Error> {
-    match std::fs::read(path) {
-        Ok(cert_bytes) => Ok(Some(cert_bytes)),
-        Err(ref err) if err.kind() == std::io::ErrorKind::NotFound => Ok(None),
-        Err(err) => Err(Error::Internal(InternalError::ReadFile(err))),
+fn build_name(subj: &CertSubject) -> Result<X509Name, Error> {
+    let mut builder = X509Name::builder()
+        .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+
+    match subj {
+        CertSubject::CommonName(cn) =>
+            builder
+                .append_entry_by_text("CN", &cn)
+                .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?,
+        CertSubject::Subject(fields) =>
+            for (name, value) in fields {
+                builder
+                    .append_entry_by_text(name, value)
+                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+            }
     }
+
+    Ok(builder.build())
 }
 
-fn create_cert<'a>(
+fn create_cert_inner<'a>(
     api: &'a mut Api,
     id: &'a str,
     csr: &'a [u8],
-    issuer: Option<(&'a str, &'a aziot_key_common::KeyHandle)>,
-) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<Vec<u8>, Error>> + Send + 'a>> {
+    issuer: Option<(&'a str, &'a KeyHandle)>
+) -> Pin<Box<dyn Future<Output = Result<Vec<u8>, BoxedError>> + Send + 'a>> {
```

Comment:
Why is this returning `Box<dyn std::error::Error>` instead of our `Error` ?

Reply:
Most errors that were not previously marked as invalid parameter errors were being wrapped in `Error::Internal(InternalError::CreateCert(Box::new(err)))`, so I made the inner function return a boxed error that we could later wrap in `Api::create_cert` using the same method.